### PR TITLE
[dbus] introspect xml

### PR DESCRIPTION
**Pull request by gjc13**
This change contains a preview of Thread network management d-bus api.

### CODE REVIEW

File: src/dbus/instropect.xml [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+ "http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd">
+<!-- wpan name will vary -->
+<node name="/org/otbr/wpan%s">
+  <interface name="org.Scan">
+    <method name="Scan">
+      <!-- ret name=scan_result type=a{r{ta{b}ta{b}qqbbbbb}} -->
+    </method>
+
+    <method name="Attach">
+      <arg name="network_name" type="s"/>
```
Comment (bukepo) [inline src/dbus/instropect.xml]: We want to make this a formal API definition. To reduce confusion, I'd like to keep the same names across all APIs. I think names used in [OpenThread CLI](https://github.com/openthread/openthread/blob/master/src/cli/README.md) are good candidates, so I tend to use `networkname` here.  What do you think? @jwhui @librasungirl ?

File: src/dbus/instropect.xml [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+      <arg name="masterkey" type="a{b}" direction="out"/>
+      <!-- arg name=channel type=a{b} -->
+    </method>
+
+    <method name="GetCcaFailureRate">
+      <arg name="failure_rate" type="q" direction="out"/>
+    </method>
+
+    <method name="GetLinkCounters">
+      <arg name="link_counters" type="v{uuuuuuuuuuuuuuuuuuuuuuuuuuuuuuu}" direction="out"/>
```
Comment (bukepo) [inline src/dbus/instropect.xml]: use concrete counter names instead of a vector of `u`?

File: src/dbus/instropect.xml [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    </method>
+
+    <method name="GetIp6Counters">
+      <arg name="link_counters" type="v{uuuu}" direction="out"/>
```
Comment (bukepo) [inline src/dbus/instropect.xml]: use concrete counter names.

File: src/dbus/instropect.xml [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+      <arg name="has_network_data" type="b"/>
+      <arg name="err" type="b" direction="out"/>
+    </method>
+
+    <method name="SetMasterKey">
+      <arg name="master_key" type="a{b}"/>
+      <arg name="err" type="b" direction="out"/>
+    </method>
+
+    <method name="SetPSKc">
```
Comment (bukepo) [inline src/dbus/instropect.xml]: I tend to define these as DBus properties, which will have some standard supporting tools.

File: src/dbus/instropect.xml [hunk 5]
```diff
@@ -0,0 +1,163 @@
+<!DOCTYPE node PUBLIC "-//freedesktop//DTD D-BUS Object Introspection 1.0//EN"
+ "http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd">
+<!-- wpan name will vary -->
+<node name="/org/otbr/wpan%s">
+  <interface name="org.Scan">
```
Comment (bukepo) [inline src/dbus/instropect.xml]: Add standard interfaces:
* `org.freedesktop.DBus.Introspectable`
* `org.freedesktop.DBus.Properties`
* `org.freedesktop.DBus.ObjectManager `

File: src/dbus/instropect.xml [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+    <method name="Scan">
+      <arg name="scan_result" type="a{r{ta{b}ta{b}qqbbbbb}}" direction="out"/> 
+    </method>
+
+    <method name="Attach">
+      <arg name="networkname" type="s"/>
+      <arg name="panid" type="q"/>
+      <arg name="extpanid" type="t"/>
+      <arg name="master_key" type="a{b}"/>
+      <arg name="channel" type="b"/>
```
Comment (bukepo) [inline src/dbus/instropect.xml]: remove this one.

File: src/dbus/instropect.xml [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+
+    <property name="LinkMode" type="(bbbb)" access="readwrite">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="MasterKey" type="ab" access="readwrite">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="PSKc" type="ab" access="write">
```
Comment (bukepo) [inline src/dbus/instropect.xml]:
```suggestion
```

File: src/dbus/instropect.xml [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+
+    <method name="SetLegacyULAPrefix">
+      <arg name="prefix" type="a{b}"/>
+    </method>
+
+    <property name="LinkMode" type="(bbbb)" access="readwrite">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="MasterKey" type="ab" access="readwrite">
```
Comment (bukepo) [inline src/dbus/instropect.xml]:
```suggestion
    <property name="MasterKey" type="ab" access="read">
```

File: src/dbus/instropect.xml [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="PSKc" type="ab" access="write">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="DeviceRole" type="b" access="read">
```
Comment (bukepo) [inline src/dbus/instropect.xml]:
```suggestion
    <property name="DeviceRole" type="string" access="read">
```

File: src/dbus/instropect.xml [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+
+    <property name="PanId" type="q" access="read">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="ExtPanId" type="t" access="read">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="Channel" type="b" access="read">
```
Comment (bukepo) [inline src/dbus/instropect.xml]:
```suggestion
    <property name="Channel" type="q" access="read">
```

File: src/dbus/instropect.xml [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+
+    <property name="LinkRxFailure" type="u" access="read">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="LinkSupportedChannelMask" type="u" access="read">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="RadioSupportedChannelMask" type="u" access="read">
```
Comment (bukepo) [inline src/dbus/instropect.xml]:
```suggestion
    <property name="SupportedChannelMask" type="u" access="read">
```

File: src/dbus/instropect.xml [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+
+    <property name="CcaFailureRate" type="q" access="read">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="MacTxTotal" type="u" access="read">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="MacTxUnicast" type="u" access="read">
```
Comment (bukepo) [inline src/dbus/instropect.xml]: aggregate into a struct.

File: src/dbus/instropect.xml [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+
+    <property name="LinkTxFailure" type="u" access="read">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="LinkRxFailure" type="u" access="read">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="LinkSupportedChannelMask" type="u" access="read">
```
Comment (bukepo) [inline src/dbus/instropect.xml]:
```suggestion
```

File: src/dbus/instropect.xml [hunk 14]
```diff
@@ -0,0 +1,131 @@
+<!DOCTYPE node PUBLIC "-//freedesktop//DTD D-BUS Object Introspection 1.0//EN"
+ "http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd">
+<node>
+  <interface name="io.openthread.borderrouter">
```
Comment (bukepo) [inline src/dbus/instropect.xml]:
```suggestion
  <interface name="io.openthread.BorderRouter">
```

File: src/dbus/instropect.xml [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+    <property name="LegacyULAPrefix" type="ab" access="readwrite">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="LinkMode" type="(bbbb)" access="readwrite">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="DeviceRole" type="string" access="read">
```
Comment (bukepo) [inline src/dbus/instropect.xml]:
```suggestion
    <property name="DeviceRole" type="s" access="read">
```

File: src/dbus/instropect.xml [hunk 16]
```diff
@@ -0,0 +1,131 @@
+<!DOCTYPE node PUBLIC "-//freedesktop//DTD D-BUS Object Introspection 1.0//EN"
+ "http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd">
+<node>
+  <interface name="io.openthread.BordeRrouter">
```
Comment (librasungirl) [inline src/dbus/instropect.xml]: nit: BorderRouter

File: src/dbus/instropect.xml [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+    <method name="Scan">
```
Comment (librasungirl) [inline src/dbus/instropect.xml]: do we want to scan specific channel?

File: src/dbus/instropect.xml [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+      <arg name="scan_result" type="a(taytayqqyyyyy)" direction="out"/>
```
Comment (librasungirl) [inline src/dbus/instropect.xml]: where do we want to host comment to explain the meaning of results returned?

Comment (librasungirl) [reply src/dbus/instropect.xml]: How about follow comment style as in spinel.h?
e.g. https://github.com/openthread/openthread/blob/master/src/ncp/spinel.h#L1645

File: src/dbus/instropect.xml (hunk 17 above)
Comment (gjc13) [reply src/dbus/instropect.xml]: We are not using this feature for the time being. Can be added afterwards.

File: src/dbus/instropect.xml [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+
+    <property name="MeshLocalPrefix" type="ay" access="readwrite">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="LegacyULAPrefix" type="ay" access="readwrite">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="LinkMode" type="(yyyy)" access="readwrite">
```
[the remark this answers is not in the document]
Comment (gjc13) [reply src/dbus/instropect.xml]: Replying boolean in dbus is tricky in code implementation so we use uint8 instead.

File: src/dbus/instropect.xml [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+
+    <property name="NetworkName" type="s" access="read">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="PanId" type="q" access="read">
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <property name="ExtPanId" type="t" access="read">
```
[the remark this answers is not in the document]
Comment (bukepo) [reply src/dbus/instropect.xml]: @jwhui is an `unsigned 64-bit integer` acceptable?

[the remark this answers is not in the document]
Comment (gjc13) [reply src/dbus/instropect.xml]: Pro of an integer is that it's simple and it's exactly how the current clients are treating the extpanid field.
Con is that it requires us to treat the endianness carefully.

[the remark this answers is not in the document]
Comment (bukepo) [reply src/dbus/instropect.xml]: @gjc13 I think you mean some internal code?

From
https://github.com/openthread/wpantund/blob/master/src/wpanctl/tool-cmd-form.c#L144
https://github.com/openthread/wpantund/blob/master/src/wpanctl/tool-cmd-form.c#L335
wpantund treats it as a uint64_t as well.

@jwhui could you give us some guidance here, please?

[the remark this answers is not in the document]
Comment (gjc13) [reply src/dbus/instropect.xml]: We've decided to keep the uint64 data format.

[the remark this answers is not in the document]
Comment (jwhui) [reply src/dbus/instropect.xml]: 64-bit unsigned integer is fine with me. We are free to represent an 8-byte value in whatever way is most convenient.

File: src/dbus/instropect.xml [hunk 21]
```diff
@@ -0,0 +1,214 @@
...
+        uint32 rx_duplicated;
+        uint32 rx_err_no_frame;
+        uint32 rx_err_unknown_neighbor;
+        uint32 rx_err_invalid_src_addr;
+        uint32 rx_err_sec;
+        uint32 rx_err_fcs;
+        uint32 rx_err_other;
+      }
+    -->
+    <property name="MacCounters" type="(uuuuuuuuuuuuuuuuuuuuuuuuuuuuuuu)" access="read">
```
Comment (librasungirl) [inline src/dbus/instropect.xml]: here 31x `u`, however 32 fields in the structure.

File: src/dbus/instropect.xml [hunk 22]
```diff
@@ -0,0 +1,214 @@
...
+      <annotation name="org.freedesktop.DBus.Property.EmitsChangedSignal" value="false"/>
+    </property>
+
+    <!--
```
Comment (librasungirl) [inline src/dbus/instropect.xml]: add comment to introduce what these counters are?

**Comment by gjc13** [comment]
@wgtdkp @Moandor-y Any more comments?

**Comment by bukepo** [comment]
@abtink @jwhui any comments?

**Pull request merged**